File: python/factory.py
# Product interface
class Animal:
    def speak(self):
        pass

# Concrete Products
class Dog(Animal):
    def speak(self):
        return "Woof!"

class Cat(Animal):
    def speak(self):
        return "Meow!"

# Factory
class AnimalFactory:
    def create_animal(self, animal_type):
        if animal_type.lower() == 'dog':
            return Dog()
        elif animal_type.lower() == 'cat':
            return Cat()
        else:
            raise ValueError("Invalid animal type")

# Client code
factory = AnimalFactory()

dog = factory.create_animal('dog')
cat = factory.create_animal('cat')

print(dog.speak())  # Output: Woof!
print(cat.speak())  # Output: Meow!

File: go/factory.go
package factory

import "fmt"

// Product interface
type Animal interface {
	Speak() string
}

// Concrete Products
type Dog struct{}

func (d *Dog) Speak() string {
	return "Woof!"
}

type Cat struct{}

func (c *Cat) Speak() string {
	return "Meow!"
}

// Factory interface
type AnimalFactory interface {
	CreateAnimal() Animal
}

// Concrete Factories
type DogFactory struct{}

func (df *DogFactory) CreateAnimal() Animal {
	return &Dog{}
}

type CatFactory struct{}

func (cf *CatFactory) CreateAnimal() Animal {
	return &Cat{}
}

// Client code
func main() {
	dogFactory := &DogFactory{}
	catFactory := &CatFactory{}

	dog := dogFactory.CreateAnimal()
	cat := catFactory.CreateAnimal()

	fmt.Println(dog.Speak()) // Output: Woof!
	fmt.Println(cat.Speak()) // Output: Meow!
}
